File: term-bench2/retry_provider.py
#!/usr/bin/env python3
"""
retry_provider.py — re-run a runner.py subcommand until the model provider is
healthy, backing off between attempts.

Why: when the upstream model provider is erroring (opencode emits an error event
and does no work — every run reports `turns=0`), a `run`/`ab` produces a
meaningless all-fail result. This wrapper detects that state and retries instead
of accepting the empty result.

Health signal: the provider is "down" only when opencode logged a transient
provider error AND nothing in the attempt did real work — i.e. every run failed
fast before doing anything. A run that produced turns, hit the agent timeout, or
passed all mean the provider is up (even if the task itself failed), so we stop.
This deliberately treats a *timeout* as "up" — it is a slow/stuck task, not a
provider outage, and retrying it would loop forever.

Backoff: the interval doubles each failed attempt and is capped at 10 minutes
(600 s): 30 → 60 → 120 → 240 → 480 → 600 → 600 …

Usage:
  python3 retry_provider.py [--base SEC] [--cap SEC] [--max-attempts N] -- <runner args...>

Examples:
  python3 retry_provider.py -- ab --layer account-global --candidate v1 \
      --tasks code-from-image constraints-scheduling --k 1
  python3 retry_provider.py --cap 600 -- run --task-file baseline-tasks.txt --k 1
"""
from __future__ import annotations

import argparse
import re
import subprocess
import sys
import time
from pathlib import Path

SCRIPT_DIR = Path(__file__).resolve().parent
RUNNER = SCRIPT_DIR / "runner.py"

# Evidence the provider actually did something this attempt: real turns, a task
# that hit the agent timeout (ran but slow/stuck), or an outright pass.
REALWORK_RE = re.compile(r"turns=[1-9]\d*|opencode timed out after|reward=1")
# run_opencode logs this line each time it detects a transient provider error.
TRANSIENT_MARK = "transient provider error"


def main() -> None:
    ap = argparse.ArgumentParser(
        description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--base", type=float, default=30.0,
                    help="initial backoff seconds (default 30)")
    ap.add_argument("--cap", type=float, default=600.0,
                    help="max backoff seconds; the interval attenuates up to this "
                         "cap (default 600 = 10 min)")
    ap.add_argument("--max-attempts", type=int, default=0,
                    help="give up after N provider-down attempts (0 = unlimited)")
    ap.add_argument("rest", nargs=argparse.REMAINDER,
                    help="-- then the runner.py subcommand + args")
    args = ap.parse_args()

    runner_args = args.rest
    if runner_args and runner_args[0] == "--":
        runner_args = runner_args[1:]
    if not runner_args:
        sys.exit("retry_provider: provide runner.py args after --")

    attempt = 0
    delay = args.base
    while True:
        attempt += 1
        print(f"\n=== retry_provider attempt {attempt}: python3 runner.py "
              f"{' '.join(runner_args)} ===", flush=True)

        proc = subprocess.Popen(
            [sys.executable, str(RUNNER), *runner_args],
            stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True)
        assert proc.stdout is not None
        saw_transient = False
        saw_realwork = False
        for line in proc.stdout:
            sys.stdout.write(line)
            sys.stdout.flush()
            if TRANSIENT_MARK in line:
                saw_transient = True
            if REALWORK_RE.search(line):
                saw_realwork = True
        proc.wait()

        # Down only if the provider errored on every run and nothing ran.
        provider_down = saw_transient and not saw_realwork
        if not provider_down:
            print(f"=== provider responding (real work seen / no transient outage) — "
                  f"done, exit={proc.returncode} ===", flush=True)
            sys.exit(proc.returncode)

        if args.max_attempts and attempt >= args.max_attempts:
            print(f"=== gave up: provider still down after {attempt} attempt(s) ===",
                  flush=True)
            sys.exit(2)

        print(f"=== provider down (all runs failed fast with transient errors) — "
              f"backing off {delay:.0f}s before attempt {attempt + 1} ===", flush=True)
        time.sleep(delay)
        delay = min(delay * 2, args.cap)


if __name__ == "__main__":
    main()

File: opencode-plugin/src/bench/retry-provider.ts
/**
 * retry-provider.ts — re-run `bun term-bench2/runner.ts <args>` until the
 * model provider is healthy, backing off between attempts. Full port of
 * term-bench2/retry_provider.py (deleted in this commit — this file plus
 * the term-bench2/retry-provider.ts shim are now the only implementation).
 *
 * Why: when the upstream model provider is erroring (opencode emits an
 * error event and does no work — every run reports `turns=0`), a `run`/`ab`
 * produces a meaningless all-fail result. This wrapper detects that state
 * and retries instead of accepting the empty result.
 *
 * Health signal: the provider is "down" only when opencode logged a
 * transient provider error AND nothing in the attempt did real work — i.e.
 * every run failed fast before doing anything. A run that produced turns,
 * hit the agent timeout, or passed all mean the provider is up (even if the
 * task itself failed), so we stop. This deliberately treats a *timeout* as
 * "up" — it is a slow/stuck task, not a provider outage, and retrying it
 * would loop forever.
 *
 * Backoff: the interval doubles each failed attempt and is capped at 10
 * minutes (600s) by default: 30 -> 60 -> 120 -> 240 -> 480 -> 600 -> 600 ...
 *
 * Marker strings: REALWORK_RE / TRANSIENT_MARK are imported (not
 * re-declared) from agent-run.ts, the PRODUCER of the log lines this
 * scans — see that file's header for the shared-constant contract.
 */
import { join } from "node:path"
import { makeBenchPaths } from "./paths.ts"
import { REALWORK_RE, TRANSIENT_MARK } from "./agent-run.ts"
import { pyFixed } from "./util.ts"

// See exec.ts's header note on why Bun globals are declared locally instead
// of depending on `bun-types` (no new deps).
declare const Bun: {
  spawn(
    cmd: string[],
    opts: { stdout: "pipe"; stderr: "pipe"; env?: Record<string, string | undefined> },
  ): {
    readonly stdout: ReadableStream<Uint8Array>
    readonly exited: Promise<number>
  }
}

export interface RetryProviderArgs {
  base: number
  cap: number
  maxAttempts: number
  runnerArgs: string[]
}

/**
 * `[--base SEC] [--cap SEC] [--max-attempts N] -- <runner.ts args...>` —
 * mirrors retry_provider.py's argparse (--base default 30, --cap default
 * 600, --max-attempts default 0 = unlimited). Returns null on any parse
 * error or an empty runner-args tail.
 */
export function parseRetryProviderArgs(argv: string[]): RetryProviderArgs | null {
  let base = 30
  let cap = 600
  let maxAttempts = 0
  let i = 0
  while (i < argv.length) {
    const a = argv[i]
    if (a === "--base" || a === "--cap" || a === "--max-attempts") {
      const v = argv[i + 1]
      if (v === undefined) return null
      const n = Number(v)
      if (Number.isNaN(n)) return null
      if (a === "--base") base = n
      else if (a === "--cap") cap = n
      else maxAttempts = n
      i += 2
      continue
    }
    break
  }
  let rest = argv.slice(i)
  if (rest[0] === "--") rest = rest.slice(1)
  if (rest.length === 0) return null
  return { base, cap, maxAttempts, runnerArgs: rest }
}

export type SleepFn = (seconds: number) => Promise<void>
async function defaultSleep(seconds: number): Promise<void> {
  await new Promise((resolve) => setTimeout(resolve, seconds * 1000))
}

/** One attempt's line-merged stdout+stderr stream, plus its exit code. */
export interface SpawnedRun {
  lines: AsyncIterable<string>
  exited: Promise<number>
}

export type SpawnFn = (runnerArgs: string[]) => SpawnedRun

/**
 * Spawn `bun term-bench2/runner.ts <runnerArgs>`, merging stderr into
 * stdout AT THE OS LEVEL (`2>&1` inside a `bash -c` wrapper) so the merged
 * stream preserves true interleaved ordering — matching Python's
 * `subprocess.Popen(..., stderr=subprocess.STDOUT)`, which Bun.spawn has no
 * direct equivalent for (its stdout/stderr are always two separate pipes).
 */
function defaultSpawn(runnerArgs: string[]): SpawnedRun {
  const runnerPath = join(makeBenchPaths().metaRoot, "term-bench2", "runner.ts")
  const proc = Bun.spawn(["bash", "-c", 'exec "$0" "$@" 2>&1', "bun", runnerPath, ...runnerArgs], {
    stdout: "pipe",
    stderr: "pipe",
    env: process.env,
  })

  async function* lineGen(): AsyncGenerator<string> {
    const reader = proc.stdout.getReader()
    const decoder = new TextDecoder()
    let buf = ""
    for (;;) {
      const { done, value } = await reader.read()
      if (done) break
      buf += decoder.decode(value, { stream: true })
      let idx: number
      while ((idx = buf.indexOf("\n")) !== -1) {
        yield buf.slice(0, idx)
        buf = buf.slice(idx + 1)
      }
    }
    if (buf) yield buf
  }

  return { lines: lineGen(), exited: proc.exited }
}

/**
 * The retry loop itself — port of retry_provider.py:68-105's `main` body.
 * `writeLine` defaults to echoing every merged line to the real stdout (so a
 * human watching the wrapper sees everything the wrapped runner printed);
 * injectable so tests can capture output instead.
 */
export async function runRetryProvider(
  args: RetryProviderArgs,
  spawnFn: SpawnFn = defaultSpawn,
  sleepFn: SleepFn = defaultSleep,
  writeLine: (line: string) => void = (l) => console.log(l),
): Promise<number> {
  let attempt = 0
  let delay = args.base
  for (;;) {
    attempt += 1
    writeLine(`\n=== retry_provider attempt ${attempt}: bun term-bench2/runner.ts ${args.runnerArgs.join(" ")} ===`)

    const { lines, exited } = spawnFn(args.runnerArgs)
    let sawTransient = false
    let sawRealwork = false
    for await (const line of lines) {
      writeLine(line)
      if (line.includes(TRANSIENT_MARK)) sawTransient = true
      if (REALWORK_RE.test(line)) sawRealwork = true
    }
    const rc = await exited

    // Down only if the provider errored on every run and nothing ran.
    const providerDown = sawTransient && !sawRealwork
    if (!providerDown) {
      writeLine(`=== provider responding (real work seen / no transient outage) — done, exit=${rc} ===`)
      return rc
    }

    if (args.maxAttempts && attempt >= args.maxAttempts) {
      writeLine(`=== gave up: provider still down after ${attempt} attempt(s) ===`)
      return 2
    }

    writeLine(
      `=== provider down (all runs failed fast with transient errors) — backing off ${pyFixed(delay, 0)}s before attempt ${attempt + 1} ===`,
    )
    await sleepFn(delay)
    delay = Math.min(delay * 2, args.cap)
  }
}

export async function main(argv: string[]): Promise<number> {
  const args = parseRetryProviderArgs(argv)
  if (args === null) {
    console.error("retry_provider: provide runner.ts args after --")
    return 1
  }
  return runRetryProvider(args)
}
